### ISSUE TITLE
no-descending-specificity false positive

### ISSUE THREAD

**Issue by mrchief** [stylelint/stylelint#3135]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A Bug where using grouped selectors with PostCSS/CSSNext, stylelint throws this error despite the CSS desugaring correctly.

> Which rule, if any, is this issue related to?

no-descending-specificity

> What CSS is needed to reproduce this issue?

```css
table,
.table {
  border-collapse: collapse;

  & > thead {
    border-bottom: 1px solid #ccc;
  }

  & > thead th {
    color: blue;
    text-transform: uppercase;
  }

  & > thead th,
  & > tfoot td,
  & > tbody td {
    padding: 5px;

    &:last-child,
    &:first-child {
      padding-left: 20px;
    }
  }

  & > tfoot {
    background-color: #3c3c3c;
    text-transform: uppercase;
  }
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
no specific rules, default config will suffice
```

> Which version of stylelint are you using?

8.4.0

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint \"./src/**/*.css\" --fix`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Its related to PostCSS nesting

> What did you expect to happen?

No errors to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following errors were flagged:

```shell
Expected selector "table > thead th:first-child" to come before selector ".table > thead th:last-child" (no-descending-specificity)
Expected selector "table > tfoot td:first-child" to come before selector ".table > tfoot td:last-child" (no-descending-specificity)
Expected selector "table > tfoot td:first-child" to come before selector ".table > tbody td:last-child" (no-descending-specificity)
Expected selector "table > tbody td:first-child" to come before selector ".table > tfoot td:last-child" (no-descending-specificity)
Expected selector "table > tbody td:first-child" to come before selector ".table > tbody td:last-child" (no-descending-specificity)
Expected selector "table > thead th" to come before selector ".table > thead th" (no-descending-specificity)
Expected selector "table > tbody td" to come before selector ".table > tfoot td" (no-descending-specificity)
Expected selector "table > tbody td:last-child" to come before selector ".table > tfoot td:last-child" (no-descending-specificity)
Expected selector "table > thead th:first-child" to come before selector ".table > thead th:last-child" (no-descending-specificity)
Expected selector "table > tfoot td:first-child" to come before selector ".table > tfoot td:last-child" (no-descending-specificity)
Expected selector "table > tfoot td:first-child" to come before selector ".table > tbody td:last-child" (no-descending-specificity)
Expected selector "table > tbody td:first-child" to come before selector ".table > tfoot td:last-child" (no-descending-specificity)
Expected selector "table > tbody td:first-child" to come before selector ".table > tbody td:last-child" (no-descending-specificity)
Expected selector "table > tbody td:first-child" to come before selector ".table > tfoot td:first-child" (no-descending-specificity)
```

Desugared CSS

```css
table,
.table {
  border-collapse: collapse
}
table > thead, .table > thead {
  border-bottom: 1px solid #ccc;
}
table > thead th, .table > thead th {
  color: blue;
  text-transform: uppercase;
}
table > thead th,
  table > tfoot td,
  table > tbody td,
  .table > thead th,
  .table > tfoot td,
  .table > tbody td {
  padding: 5px
}
table > thead th:last-child,
    table > thead th:first-child,
    table > tfoot td:last-child,
    table > tfoot td:first-child,
    table > tbody td:last-child,
    table > tbody td:first-child,
    .table > thead th:last-child,
    .table > thead th:first-child,
    .table > tfoot td:last-child,
    .table > tfoot td:first-child,
    .table > tbody td:last-child,
    .table > tbody td:first-child {
  padding-left: 20px;
}
table > tfoot, .table > tfoot {
  background-color: #3c3c3c;
  text-transform: uppercase;
}
```

[Codepen](https://codepen.io/mrchief/pen/EoGrdO)

Errors repro using demo site:

![image](https://user-images.githubusercontent.com/781818/35166071-94be3572-fd1e-11e7-8390-4e8e544a5fda.png)

**Comment by jeddy3** [comment]
@mrchief Thanks for the detailed report and for using the template.

I believe the rule is behaving [as documented](https://github.com/stylelint/stylelint/blob/master/lib/rules/no-descending-specificity/README.md). It compares the desugared CSS e.g. `table > thead th:first-child` has a specificity of `0,1,3` whereas `.table > thead th:last-child` has a specificity of `0,2,2`. So the source order and the specificity order are not aligned. The desugared CSS has interwoven class selectors and element selectors. The rule is designed to warn against these. 

I can think of two options, either refactor so they aren't interwoven or turn off the rules for this code block using `/* stylelint-disable .. */` command comments.

**Comment by mrchief** [comment]
@jeddy3 Thanks for your response and explanation. I totally get your point. However, this feels like a little bizarre in this scenario. And I understand that static analysis can only do so much. `.table` class exists only for enabling table like styling to div based tabular layouts. While `table` and `.table` will never exist together, the linter has no possible way of knowing that. But if one were to mix-n-match, there will be surprises so I totally understand what the linter is trying to prevent.

Refactoring this will remove the DRYness or make it less legible which I think are bigger maintenance nigthmares. 

For now, I'll disable for this block. 

The only suggestion I would have at this point is to improve the error message (if possible). What you said, about specificity, is a much clearer message - and I think - had that been in the error - I wouldn't have raised the issue in the first place! 😃  
I got sidetracked by the existing message since it seems misguided and didn't even think about specificity. 

I think this issue is resolved for me at this point, but since you tagged it as discussion, I'll leave it up to you to close or continue this. 

Thanks again for the clear explanation!

**Comment by s3w47m88** [comment]
I experience this as well. It seems nonsensical from my perspective. It's extremely difficult to trace back where the original issue occurs. I would really like to see an option to disable this or switch it to disabled by default.
